pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -204,6 +222,48 @@ func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconci
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileNodeMaintenance) checkQuorumValidity(node *corev1.Node) (bool, error) {
+
+	if !isMasterNode(node) {
+		log.Infof("node %s is not a master", node.Name)
+		return true, nil
+	}
+
+	masterNodes, availableMasterNodes, err := r.countMasterNodes()
+	if err != nil {
+		return false, err
+	}
+
+	quorumValid := availableMasterNodes-1 >= (masterNodes/2 + 1)
+
+	log.Infof("Node %s is a master, but quorum is still maintained. (availableMasters %d totalMasters %d", node.Name, availableMasterNodes, masterNodes)
+
+	return quorumValid, nil
+}
+
+func (r *ReconcileNodeMaintenance) countMasterNodes() (int, int, error) {
+	masterNodes := 0
+	availableMasterNodes := 0
+
+	n, err := r.drainer.Client.CoreV1().Nodes().List(metav1.ListOptions{LabelSelector: MasterNodeLabelSelector})
+	if err != nil {
+		return -1, -1, err
+	}
+	for _, node := range n.Items {
+		if !node.Spec.Unschedulable {
+			availableMasterNodes += 1
```

Comment:
An admin might mark a node as unschedulable. This doesn't mean it's not available for quorum

Reply:
That is generally true, but for this exercise I cant tell if the node has been marked as unschedulable for the purpose of takeing it into maintenance (i.e. to shut it down) or not.

Reply:
What about adding an annotation to the node when it's being unschedulable by NMO?

Reply:
I don't know if you want to do that: you get race conditions if a node is being set for maintenance and hasn't reached that stage yet.

Reply:
After discussion with @beekhof  - he says we should rather err on the side of caution, so leave it as it is now.

Reply:
now we are looking at taints instead of the unschedulable property , which is more reliable.

Reply:
now using pdb based solution instead.